hypothesis-python/tests/cover/test_lookup.py

```diff
@@ -655,3 +656,26 @@ class TreeForwardRefs(typing.NamedTuple):
 @given(st.builds(TreeForwardRefs))
 def test_resolves_forward_references_outside_annotations(t):
     assert isinstance(t, TreeForwardRefs)
+
+
+def constructor(a: str = None):
+    pass
+
+
+class WithOptionalInSignature:
+    __signature__ = inspect.signature(constructor)
+    __annotations__ = typing.get_type_hints(constructor)
+
+    def __init__(self, **kwargs):
+        assert set(kwargs) == {"a"}
+        self.a = kwargs["a"]
+
+
+def test_compat_get_type_hints_aware_of_None_default():
+    # Regression test for https://github.com/HypothesisWorks/hypothesis/issues/2648
+    strategy = st.builds(WithOptionalInSignature, a=infer)
```

Comment:
What about the same logic, but:
- for `constructor` itself?
- for `WithOptionalInSignature` without `infer`?

Reply:
> What about the same logic, but:
> 
> * for `constructor` itself?

Works via `typing.get_type_hints`, and we don't do anything fancy because it's not a class with a `__signature__` attribute.

> * for `WithOptionalInSignature` without `infer`?

`st.builds()` doesn't infer strategies for non-required arguments, so this will always get the default argument of `None`.